### ISSUE TITLE
FLP integration: Back navigation with routing

### ISSUE THREAD

**Issue by christianguenter2** [abap2UI5/abap2UI5#1420]
We like to use FLP back navigation to navigate back inside our abap2UI5 apps. Currently back navigation always returns to FLP start page. AFAIS we don't have means to do that? 

Standard app in detail view, FLP back button returns to overview
![image](https://github.com/user-attachments/assets/3df378fa-9e77-4977-912c-9c510917d72e)
![image](https://github.com/user-attachments/assets/05d71f40-1cb0-4f97-b540-7e1bff220867)

If I'm not mistaken it's done via routing. Any idea whether we can do something similar?

**Comment by oblomov-dev** [comment]
yes so far just navigation between apps is possible. inside an app we have to do routing and then the FLP remembers it when pressing back. Its an important functionality, which would be nice becoming part of abap2UI5 :)

not sure if we can just use the JS history API or if we have to use the UI5 routing which demands different views as far as i remember. that would be a bit more tricky.

maybe you can share as a first step a basic sample which only contains the needed functionality to test routing?

as a second step you can add a new frontend function (as it is available for the normal flp navigation) and play around with the history API of in JS. maybe you have luck. or someone else here finds a way?

i will come also back to this, but need to finish some other things before.

**Comment by axelmohnen** [comment]
Hi
I found this interesting blog about the topic "Restore fiori app state":
[https://community.sap.com/t5/technology-blogs-by-members/restoring-fiori-application-state/ba-p/13319379](url)

Table stores the app state: 
/UIF/LREPDCONTCD

Handler Classes:
/UIF/CL_LREP_API
/UIF/CL_LREP_DB_ACCESS_CDEP

Maybe we could store the whole 2UI5 "app stringify" before calling the cross application navigation.

**Comment by oblomov-dev** [comment]
thanks for sharing this blog post, very interesting!

the good thing is, we dont need to extra stringify anything, because abap2UI5 already keeps a draft of the app after navigation, thats why the backend navigation works, eg:
client->nav_app_leave( )

here in the frontend situation, they use an additional url parameter to save the identifier:
<img width="800" alt="Screenshot 2024-12-12 at 00 47 37" src="https://github.com/user-attachments/assets/70f3f5d6-ad9d-42b4-bbd8-d779c3559c91" />

played around with this once, think something similar could also work for the actual frontend navigation of abap2UI5. will check that and come back.

**Comment by oblomov-dev** [comment]
app state feature now integrated into abap2UI5, see sample z2ui5_cl_demo_app_321.

links:
https://community.sap.com/t5/technology-blogs-by-members/restoring-fiori-application-state/ba-p/13319379
https://community.sap.com/t5/technology-q-a/creating-appstate-in-fiori-launchpad/qaq-p/11902597
https://embed.plnkr.co/wp6yes?show=view%2FApp.view.xml,preview

**Comment by axelmohnen** [comment]
Hi @oblomov-dev,
z2ui5_cl_demo_app_321 works perfectly!

I guess for the FLP cross application navigation "client->cs_event-cross_app_nav_to_ext" we 
have to call method "client->set_app_state_active( )" just before?

Thanks,
Axel

**Comment by oblomov-dev** [comment]
Yes never tried but this is the idea :) feel free to to test it and let us know if it works

**Comment by oblomov-dev** [comment]
if someone wants to try this functionality.

First approach for navigation via browser back/forward is in the project, try samples:
z2ui5_cl_demo_app_468
Z2UI5_CL_DEMO_APP_480

And press in the browser back, forward.
480 keeps addiitonally the state.

Feedback welcome!

**Comment by oblomov-dev** [comment]
close this for now, the functionality is implemted. if someone runs in problems or needs new/different features. just open a new issue.

actual change:
In-app back for the FLP/browser back button. An app can own its URL hash 1:1 like a UI5 router: register cs_event-hash_attach_changed, write routes with client->hash_set( '/detail/0' ) — a real, pushed history entry through the UI5 HashChanger (the FLP shell hash is preserved) — rewrite in place with hash_replace( ), and step back with cs_event-hash_back: one consumed window.history.go(-1) with an optional fallback hash for cold deep links (the UI5 onNavBack pattern). Because every in-app navigation is now a real history entry, the FLP back button pops the app's own entry and navigates back inside the app instead of falling out to the launchpad home page. Alternatively, draft-based routing via cs_event-hash_routing (KEEP/FRESH) restores whole app states on Back/Forward, and app_state_set_active( ) + the new app_state_get_href( ) cover bookmarks and share links (FLP-safe, composed backend-side).

Samples, all in the new Hash category: z2ui5_cl_smp_app_499 (the whole hash family in one app), 468/480 (routing modes fresh/keep), 498 (app state, bookmark & share). Full router emulations of demo-kit apps live in samples-controls: z2ui5_cl_smpc_app_012 and the FlexibleColumnLayout family 575/577/578/579/580/584.
